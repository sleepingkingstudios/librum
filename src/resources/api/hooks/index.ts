export { useResourceQuery } from './use-resource-query';
export type { UseResourceQueryOptions } from './use-resource-query';
