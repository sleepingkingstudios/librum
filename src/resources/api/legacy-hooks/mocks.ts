import { defaultResponse } from '@api/test-helpers';
import type { UseResourceQuery } from '../types';

const resourceQuery: jest.MockedFunction<UseResourceQuery> =
  jest.fn(() => defaultResponse);

export const useResourceQuery: jest.MockedFunction<() => UseResourceQuery> =
  jest.fn(() => resourceQuery);
