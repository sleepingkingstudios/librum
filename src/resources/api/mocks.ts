import type { UseQuery } from '@api';
import { defaultResult } from '@api/test-helpers';
import type {
  ResourceApiEndpointConfiguration,
  ResourceConfiguration,
} from '../types';
import type { ResourceApiHooks } from './types';

const useIndexResources: jest.MockedFunction<UseQuery> =
  jest.fn(() => defaultResult);

type GenerateResourcesApiProps = {
  endpoints?: ResourceApiEndpointConfiguration,
} & ResourceConfiguration;

type GenerateResourcesApi =
  (options: GenerateResourcesApiProps) => ResourceApiHooks;

export const generateResourcesApi: jest.MockedFunction<GenerateResourcesApi> =
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  jest.fn((...args) => ({
    useIndexResources,
  }));
