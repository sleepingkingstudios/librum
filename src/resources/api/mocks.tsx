import { responseWithStatus } from '@api';
import type {
  Refetch,
  Response,
} from '@api';

type UseResourceQuery = (props: unknown) => [Response, Refetch];

export { generateAlerts } from './utils';

export const useResourceQuery: jest.MockedFunction<UseResourceQuery> = jest.fn(
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  (props) => [responseWithStatus({ status: 'uninitialized' }), jest.fn()]
);
