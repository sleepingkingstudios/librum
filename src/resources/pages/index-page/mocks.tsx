import * as React from 'react';

import { responseWithData } from '@api';
import type { DataTableData } from '@components/data-table';
import { ResourcePage } from '@resources/components/page';
import type { ResourcePageProps } from '@resources/components/page';

export const MockTable = ({ data }: { data: DataTableData }): JSX.Element => {
  const books = 'rareBooks' in data ? data.rareBooks : [];

  return (
    <span className="mock-data-table">
      { `There are ${books.length} books!` }
    </span>
  );
};

export const ResourceIndexPage = ({
  Table,
  page,
  resourceName,
}: ResourcePageProps): JSX.Element => {
  const data: DataTableData = {
    rareBooks: [
      {
        title: 'On War',
      },
      {
        title: 'The Art of War',
      },
      {
        title: 'The Prince',
      },
    ],
  };
  const response = responseWithData({ data });
  const contents = (
    <div>
      <h1>Content for {resourceName}</h1>

      <Table data={data} name={resourceName} />
    </div>
  );
  const pageWithDefaults = {
    ...page,
    contents,
  };

  return (
    <ResourcePage
      action={''}
      page={pageWithDefaults}
      resourceName={resourceName}
      response={response}
    />
  );
};
