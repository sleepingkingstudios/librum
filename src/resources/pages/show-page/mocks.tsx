import * as React from 'react';

import { responseWithData } from '@api';
import { DataList } from '@components/data-list';
import { ResourcePage } from '@resources/components/page';
import type { ResourcePageProps } from '@resources/components/page';
import type { DataBlockData } from '@resources/types';
import { singularize } from '@utils/inflector';
import { Literal } from '@utils/types';

const defaultBook = {
  name: 'Unknown Book',
  slug: 'unknown-book',
  author: 'Unknown Author',
};

export const MockBlock = ({ data }: { data: DataBlockData }): JSX.Element => {
  const book = (data.rareBook || defaultBook) as Record<string, Literal>;

  return (<DataList data={book} />);
};

export const ResourceShowPage = ({
  Block,
  page,
  resourceName,
}: ResourcePageProps): JSX.Element => {
  const data: DataBlockData = {
    rareBook: {
      name: 'Gideon the Ninth',
      slug: 'gideon-9',
      author: 'Tammsyn Muir',
    },
  };
  const response = responseWithData({ data });
  const contents = (
    <div>
      <h1>Content for {singularize(resourceName)}</h1>

      <Block data={data} name={resourceName} />
    </div>
  );

  const pageWithDefaults = {
    ...page,
    contents,
  };

  return (
    <ResourcePage
      action={''}
      page={pageWithDefaults}
      resourceName={resourceName}
      response={response}
    />
  );
};
