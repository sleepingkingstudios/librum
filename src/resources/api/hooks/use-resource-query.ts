import type {
  Effect,
  UseQuery,
} from '@api';
import { displayAlerts } from '@api/effects';
import { useQueryRequest } from '@api/hooks';
import type { AlertDirective } from '@api/effects/display-alerts';
import { generateAlerts } from '../effects';
import type {
  ResourceQueryParams,
  UseResourceQuery,
} from '../types';

export const useResourceQuery = ({
  action,
  alerts,
  effects,
  member,
  resourceName,
  scope,
  singularName,
  useQuery,
}: {
  action: string,
  alerts?: false | AlertDirective[],
  effects?: Effect[],
  member: boolean,
  resourceName: string,
  scope?: string,
  singularName?: string,
  useQuery: UseQuery,
}): UseResourceQuery => {
  const configuredEffects = effects ? [...effects] : [];

  if (alerts === undefined || alerts === null) {
    const configuredAlerts = generateAlerts({
      action,
      member,
      query: true,
      resourceName,
      singularName,
      scope,
    });

    configuredEffects.unshift(displayAlerts(configuredAlerts));
  } else if (alerts) {
    configuredEffects.unshift(displayAlerts(alerts));
  }

  // eslint-disable-next-line react-hooks/rules-of-hooks
  return (arg?: ResourceQueryParams) => useQueryRequest({
    arg,
    effects: configuredEffects,
    useQuery,
  });
};
