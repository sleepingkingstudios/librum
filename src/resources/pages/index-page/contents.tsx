import * as React from 'react';

import type { UseQuery } from '@api';
import { useResourceQuery } from '@resources/api/hooks';
import type { ResourcePageProps } from '@resources/components/page';
import { ResourcesTable } from '@resources/components/table';

export const ResourceIndexPageContents = ({
  Table,
  apiHooks,
  page,
  resourceName,
  scope,
  singularName,
}: ResourcePageProps): JSX.Element => {
  if (!('useIndexResources' in apiHooks)) {
    const hooks = Object.keys(apiHooks);
    let message = 'undefined API endpoint "useIndexResources"';

    if (hooks.length > 0) {
      message = `${message} - valid hooks are ${hooks.join(', ')}`;
    }

    throw new Error(`[ERROR] <ResourceIndexPageContents /> ${message}`);
  }

  if (!Table) {
    const message = 'default IndexPageContents requires Table property';

    throw new Error(`[ERROR] <ResourceIndexPageContents /> ${message}`);
  }

  const {
    alerts,
    effects,
  } = page;
  const useQuery = apiHooks.useIndexResources as UseQuery;
  const useRequest = useResourceQuery({
    ...(alerts ? { alerts } : {}),
    ...(effects ? { effects } : {}),
    ...(scope ? { scope } : {}),
    ...(singularName ? { singularName } : {}),
    action: 'index',
    member: false,
    resourceName,
    useQuery,
  });

  return (
    <ResourcesTable
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      Table={Table}
      resourceName={resourceName}
      useRequest={useRequest}
    />
  );
};
