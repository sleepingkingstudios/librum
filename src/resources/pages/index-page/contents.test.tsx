import * as React from 'react';

import '@testing-library/jest-dom';

import { ResourceIndexPageContents } from './contents';
import type {
  Effect,
  Response,
} from '@api';
import type { AlertDirective } from '@api/effects/display-alerts';
import {
  failureResponse,
  loadingResponse,
  successResponse,
} from '@api/test-helpers';
import type { DataTableData } from '@components/data-table';
import { render } from '@test-helpers/rendering';
import { useResourceQuery as mockUseResourceQuery } from '@resources/api/hooks/mocks';
import type { ResourcePageOptions } from '@resources/components/page';

jest.mock('@resources/api/hooks', () => require('@resources/api/hooks/mocks'));

const useRequest = mockUseResourceQuery() as jest.MockedFunction<() => Response>;

const MockTable = ({ data }: { data: DataTableData }): JSX.Element => {
  const books = 'rareBooks' in data ? data.rareBooks : [];

  return (
    <span className="mock-data-table">
      { `There are ${books.length} books!` }
    </span>
  );
};

describe('<ResourceIndexPageContents />', () => {
  const action = 'index';
  const member = false;
  const page: ResourcePageOptions = {};
  const resourceName = 'rareBooks';
  const useIndexResources = jest.fn();
  const apiHooks = { useIndexResources };

  beforeEach(() => {
    mockUseResourceQuery.mockClear();

    useRequest.mockClear();
  });

  it('should configure the request', () => {
    render(
      <ResourceIndexPageContents
        Table={MockTable}
        action={action}
        apiHooks={apiHooks}
        member={member}
        page={page}
        resourceName={resourceName}
      />,
      { store: true },
    );
    const expected = {
      action,
      member,
      resourceName,
      useQuery: useIndexResources,
    };

    expect(mockUseResourceQuery).toHaveBeenCalledWith(expected);
  });

  it('should perform the query', () => {
    render(
      <ResourceIndexPageContents
        Table={MockTable}
        action={action}
        apiHooks={apiHooks}
        member={member}
        page={page}
        resourceName={resourceName}
      />,
      { store: true },
    );

    expect(useRequest).toHaveBeenCalled();
  });

  it('should display the empty table', () => {
    const { getByText } = render(
      <ResourceIndexPageContents
        Table={MockTable}
        action={action}
        apiHooks={apiHooks}
        member={member}
        page={page}
        resourceName={resourceName}
      />,
      { store: true },
    );

    const table = getByText('There are 0 books!');

    expect(table).toBeVisible();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <ResourceIndexPageContents
        Table={MockTable}
        action={action}
        apiHooks={apiHooks}
        member={member}
        page={page}
        resourceName={resourceName}
      />,
      { store: true },
    );

    expect(asFragment).toMatchSnapshot();
  });

  describe('when the response is loading', () => {
    beforeEach(() => {
      useRequest.mockImplementation(() => loadingResponse);
    });

    it('should display the loading message', () => {
      const { getByText } = render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );

      const message = getByText('Loading Rare Books...');

      expect(message).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );

      expect(asFragment).toMatchSnapshot();
    });
  });

  describe('when the response is failing', () => {
    beforeEach(() => {
      useRequest.mockImplementation(() => failureResponse);
    });

    it('should display the empty table', () => {
      const { getByText } = render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );

      const table = getByText('There are 0 books!');

      expect(table).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );

      expect(asFragment).toMatchSnapshot();
    });
  });

  describe('when the response is passing', () => {
    const data = {
      rareBooks: [
        {
          title: 'On War',
        },
        {
          title: 'The Art of War',
        },
        {
          title: 'The Prince',
        },
      ],
    };
    const response = {
      ...successResponse,
      data,
      hasData: true,
    };

    beforeEach(() => {
      useRequest.mockImplementation(() => response);
    });

    it('should display the table', () => {
      const { getByText } = render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );

      const table = getByText('There are 3 books!');

      expect(table).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );

      expect(asFragment).toMatchSnapshot();
    });
  });

  describe('with page: { alerts: value }', () => {
    const alerts: AlertDirective[] = [
      {
        display: { message: 'Something went wrong' },
        status: 'failure',
      },
      {
        display: { message: 'All is well' },
        status: 'success',
      },
    ];
    const page: ResourcePageOptions = { alerts };

    it('should configure the request', () => {
      render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );
      const expected = {
        action,
        alerts,
        member,
        resourceName,
        useQuery: useIndexResources,
      };

      expect(mockUseResourceQuery).toHaveBeenCalledWith(expected);
    });
  });

  describe('with effects: value', () => {
    const effects: Effect[] = [
      jest.fn(),
      jest.fn(),
    ];
    const page: ResourcePageOptions = { effects };

    it('should configure the request', () => {
      render(
        <ResourceIndexPageContents
          Table={MockTable}
          action={action}
          apiHooks={apiHooks}
          member={member}
          page={page}
          resourceName={resourceName}
        />,
        { store: true },
      );
      const expected = {
        action,
        effects,
        member,
        resourceName,
        useQuery: useIndexResources,
      };

      expect(mockUseResourceQuery).toHaveBeenCalledWith(expected);
    });
  });
});
