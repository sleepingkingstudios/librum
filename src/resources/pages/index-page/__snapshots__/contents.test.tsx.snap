// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResourceIndexPageContents /> should match the snapshot 1`] = `[Function]`;

exports[`<ResourceIndexPageContents /> when the response is failing should match the snapshot 1`] = `[Function]`;

exports[`<ResourceIndexPageContents /> when the response is loading should match the snapshot 1`] = `[Function]`;

exports[`<ResourceIndexPageContents /> when the response is passing should match the snapshot 1`] = `[Function]`;
